### ISSUE TITLE
How to achieve deterministic results between tf2 eager and graph mode

### ISSUE THREAD

**Issue by rickstaa** [rickstaa/LAC-TF2-TORCH-translation#10]
Here a quick guide on how I achieved deterministic results that are equal between the tf2 eager and graph mode version. Below I depicted the steps needed to achieve this. To see this behaviour in action check out the[sandbox/compare_test_tf2.py](https://github.com/rickstaa/filter_LAC_tf_rewrite/blob/master/sandbox/compare_test_tf2.py) and [sandbox/compare_test_tf2_eager.py](https://github.com/rickstaa/filter_LAC_tf_rewrite/blob/master/sandbox/compare_test_tf2_eager.py) scripts.

# Set global seeds

First make sure you set the global seeds:

```python
    RANDOM_SEED = 0
    os.environ["PYTHONHASHSEED"] = str(RANDOM_SEED)
    os.environ["TF_CUDNN_DETERMINISTIC"] = "1"  # new flag present in tf 2.0+
    random.seed(RANDOM_SEED)
    np.random.seed(RANDOM_SEED)
    tf.random.set_seed(RANDOM_SEED)
    TFP_SEED_STREAM = tfp.util.SeedStream(RANDOM_SEED, salt="random_beta")
```

# Pass these seeds to the weight initializers

Because of differences between the graph and eager mode the weights are not automatically generated using the global seed. To make sure that the weights are the same you have to explicitly define your weight initializers. This can be done using the following code:

```python
initializer = tf.keras.initializers.GlorotUniform(
    seed=self._seed
)
```

This initializer can be passed to the keras api using the following code:

**Eager mode**

```pyhton
tf.keras.layers.Dense(
    hidden_size_i,
    activation="relu",
    name="l{}".format(i),
    kernel_initializer=initializer,
)
```

**Graph mode**
```python
tf.compat.v1.layers.dense(
    net_0,
    n2,
    activation=tf.nn.relu,
    name="l2",
    trainable=trainable,
    kernel_initializer=initializer,
)
```

You can also pass this initialiser to a variable using the following code:

**Eager mode**

```python
w1_s = tf.Variable(
            initializerr(shape=(self.s_dim, hidden_sizes[0])), name="w1_s",
        )
```

**Graph mode**

```python
w1_s = tf.compat.v1.get_variable(
    w1_s", [self.s_dim, n1], trainable=trainable, initializer=initializer,
)
````

# Make sure the noise sample seed is equal

Since in the Gaussian actor, we use the [sample method of the tensorflow_probaility package](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/MultivariateNormalDiag) we have to make sure the sample is sampled using the same random seed. This can be done by making use of the [tfp.util.SeedStream class](https://www.tensorflow.org/probability/api_docs/python/tfp/util/SeedStream). To do this first create a SeedStream object:

```python
TFP_SEED_STREAM = tfp.util.SeedStream(RANDOM_SEED, salt="random_beta")
```

Then create a seed using this SeedStream object and pass it to the sampling method:

**Eager mode**

```python
ga_target_seed = TFP_SEED_STREAM()
base_distribution = tfp.distributions.MultivariateNormalDiag(
    loc=tf.zeros(self.a_dim), scale_diag=tf.ones(self.a_dim)
)
epsilon = base_distribution.sample(batch_size, seed=ga_target_seed)
```

**Graph mode**

```python
ga_target_seed = TFP_SEED_STREAM()
base_distribution = tfp.distributions.MultivariateNormalDiag(
    loc=tf.zeros(self.a_dim), scale_diag=tf.ones(self.a_dim)
)
epsilon = base_distribution.sample(batch_size, seed=ga_target_seed)
```

# Create an extra copy of the Gaussian Actor in eager mode

Finally, in eager mode, we need to create an extra copy of the Gaussian Actor. This needs to be done since, in graph mode, two separate graphs are created for the main Gaussian actor and the Second Lyapunov Gaussian actor that is used in calculating the `L_delta`. In eager mode, since it is object orientated, we can simply reuse the first Gaussian actor in calculating the `L_delta`. I first tried giving this copy actor the random seed of the first actor. This however did not seem to work. I'm not 100% sure why this fails but this might be caused by the way the graph is created when `reuse==True`.

**Eager mode**

```python
self.lya_ga_target_seeds = [
    RANDOM_SEED,
    TFP_SEED_STREAM(),
]  # [weight init seed, sample seed]
base_distribution = tfp.distributions.MultivariateNormalDiag(
    loc=tf.zeros(self.a_dim), scale_diag=tf.ones(self.a_dim)
)
epsilon = base_distribution.sample(batch_size, seed=self._tfp_seed)
```

**Graph mode**

```python
lya_ga_target_seeds = [
   RANDOM_SEED,
    TFP_SEED_STREAM(),
]  # [weight init seed, sample seed]
base_distribution = tfp.distributions.MultivariateNormalDiag(
    loc=tf.zeros(self.a_dim), scale_diag=tf.ones(self.a_dim)
)
epsilon = base_distribution.sample(batch_size, seed=lya_ga_target_seeds)
````

#### Other resources
- [Tensorflow-probability documentation](https://www.tensorflow.org/probability/examples/TFP_Release_Notebook_0_11_0)
- [Medium post](https://medium.com/@ODSC/properly-setting-the-random-seed-in-ml-experiments-not-as-simple-as-you-might-imagine-219969c84752)
